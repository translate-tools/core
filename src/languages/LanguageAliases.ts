export const createLanguageAliasesMap = (languages: string[]) => {
	const complexLanguages = languages.filter((language) => language.includes('-'));

	// Build languages map
	const languagesMap: Record<string, string[]> = {};
	for (const language of complexLanguages) {
		const simpleLanguage = language.split('-')[0];

		if (!languagesMap[simpleLanguage]) {
			languagesMap[simpleLanguage] = [language];
			continue;
		}

		languagesMap[simpleLanguage].push(language);
	}

	return languagesMap;
};

export class LanguageAliases {
	private readonly simpleLanguages;
	private readonly languagesMaps;
	constructor(private readonly languagesList: string[]) {
		// Build map
		const languagesMap = createLanguageAliasesMap(languagesList);
		this.languagesMaps = {
			normal: languagesMap,
			reverse: Object.fromEntries(
				Object.entries(languagesMap)
					.map(([simpleLanguage, aliases]) =>
						aliases.map((alias) => [alias, simpleLanguage]),
					)
					.flat(),
			),
		};

		this.simpleLanguages = new Set(languagesList);
	}

	public getAll() {
		return Array.from(
			new Set([...this.languagesList, ...Object.keys(this.languagesMaps.normal)]),
		);
	}

	public get(language: string) {
		// Return mapped language
		const languageAliases = this.languagesMaps.normal[language];
		if (languageAliases) return languageAliases[0];

		// Return language in list
		if (this.simpleLanguages.has(language)) return language;

		return null;
	}
}
