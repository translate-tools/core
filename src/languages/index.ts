export * from './codes';
export * from './LanguageAliases';
