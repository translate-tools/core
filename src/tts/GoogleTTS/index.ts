import { bufferToArrayBuffer } from '../../utils/buffers';
import { Fetcher } from '../../utils/Fetcher';
import { basicFetcher } from '../../utils/Fetcher/basicFetcher';

import { TTSAudioBuffer, TTSProviderProps } from '../types';

export class GoogleTTS implements TTSProviderProps {
	private readonly fetcher;
	constructor({ fetcher = basicFetcher }: { fetcher?: Fetcher } = {}) {
		this.fetcher = fetcher;
	}

	public getAudioBuffer(text: string, language: string): Promise<TTSAudioBuffer> {
		const url =
			`https://translate.google.com/translate_tts?ie=UTF-8&tl=${language}&client=dict-chrome-ex&ttsspeed=0.5&q=` +
			encodeURIComponent(text);

		return this.fetcher(url, { responseType: 'arrayBuffer', method: 'GET' }).then(
			({ data }) => {
				let buffer: ArrayBuffer;
				if (typeof Buffer !== 'undefined' && data instanceof Buffer) {
					buffer = bufferToArrayBuffer(data);
				} else if (data instanceof ArrayBuffer) {
					buffer = data;
				} else {
					throw new TypeError('Unexpected response');
				}

				return {
					type: 'audio/mpeg',
					buffer,
				};
			},
		);
	}

	public static getSupportedLanguages() {
		// prettier-ignore
		return [
			"af", "sq", "am", "ar", "hy", "as", "ay", "az", "bm", "eu",
			"be", "bn", "bho", "bs", "bg", "ca", "ceb", "ny", "zh", "zh_HANT",
			"co", "hr", "cs", "da", "dv", "doi", "nl", "en", "eo", "et", "ee",
			"tl", "fi", "fr", "fy", "gl", "ka", "de", "el", "gn", "gu", "ht",
			"ha", "haw", "iw", "hi", "hmn", "hu", "is", "ig", "ilo", "id",
			"ga", "it", "ja", "jw", "kn", "kk", "km", "rw", "gom", "ko",
			"kri", "ku", "ckb", "ky", "lo", "la", "lv", "ln", "lt", "lg",
			"lb", "mk", "mai", "mg", "ms", "ml", "mt", "mi", "mr", "mni-Mtei",
			"lus", "mn", "my", "ne", "no", "or", "om", "ps", "fa", "pl",
			"pt", "pa", "qu", "ro", "ru", "sm", "sa", "gd", "nso", "sr",
			"st", "sn", "sd", "si", "sk", "sl", "so", "es", "su", "sw", "sv",
			"tg", "ta", "tt", "te", "th", "ti", "ts", "tr", "tk", "ak", "uk", "ur",
			"ug", "uz", "vi", "cy", "xh", "yi", "yo", "zu"
		];
	}
}
