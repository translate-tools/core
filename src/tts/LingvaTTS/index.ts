import { Fetcher } from '../../utils/Fetcher';
import { basicFetcher } from '../../utils/Fetcher/basicFetcher';

import { TTSAudioBuffer, TTSProviderProps } from '../types';

export class LingvaTTS implements TTSProviderProps {
	private readonly host;
	private readonly fetcher;
	constructor({
		fetcher = basicFetcher,
		apiHost,
	}: { fetcher?: Fetcher; apiHost?: string } = {}) {
		this.host = apiHost ?? 'https://translate.plausibility.cloud';
		this.fetcher = fetcher;
	}

	public async getAudioBuffer(text: string, language: string): Promise<TTSAudioBuffer> {
		const url = `${this.host}/api/v1/audio/${encodeURIComponent(
			language,
		)}/${encodeURIComponent(text)}`;

		return this.fetcher(url, { responseType: 'json', method: 'GET' }).then(
			({ data: json }) => {
				if (typeof json !== 'object' || json === null) {
					throw new TypeError('Unexpected response');
				}
				if (!('audio' in json) || !Array.isArray(json.audio)) {
					throw new TypeError('Unexpected response');
				}

				return {
					type: 'audio/mpeg',
					buffer: new Uint8Array(json.audio).buffer,
				};
			},
		);
	}

	public static getSupportedLanguages() {
		// prettier-ignore
		return [
			"af", "sq", "am", "ar", "hy", "as", "ay", "az", "bm", "eu",
			"be", "bn", "bho", "bs", "bg", "ca", "ceb", "ny", "zh", "zh_HANT",
			"co", "hr", "cs", "da", "dv", "doi", "nl", "en", "eo", "et", "ee",
			"tl", "fi", "fr", "fy", "gl", "ka", "de", "el", "gn", "gu", "ht",
			"ha", "haw", "iw", "hi", "hmn", "hu", "is", "ig", "ilo", "id",
			"ga", "it", "ja", "jw", "kn", "kk", "km", "rw", "gom", "ko",
			"kri", "ku", "ckb", "ky", "lo", "la", "lv", "ln", "lt", "lg",
			"lb", "mk", "mai", "mg", "ms", "ml", "mt", "mi", "mr", "mni-Mtei",
			"lus", "mn", "my", "ne", "no", "or", "om", "ps", "fa", "pl",
			"pt", "pa", "qu", "ro", "ru", "sm", "sa", "gd", "nso", "sr",
			"st", "sn", "sd", "si", "sk", "sl", "so", "es", "su", "sw", "sv",
			"tg", "ta", "tt", "te", "th", "ti", "ts", "tr", "tk", "ak", "uk", "ur",
			"ug", "uz", "vi", "cy", "xh", "yi", "yo", "zu"
		];
	}
}
