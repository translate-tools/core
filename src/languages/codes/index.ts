import languages from './ISO639LangCodesList';

/**
 * Check is string are 639-1 lang code
 *
 * Values are case sensitive, if you need, you have to convert strings to lower case to check
 */
export const isLanguageCodeISO639v1 = (code: string) => {
	return Boolean(code && languages.some((lang) => code === lang.v1));
};

/**
 * Check is string are 639-2 lang code
 *
 * Values are case sensitive, if you need, you have to convert strings to lower case to check
 */
export const isLanguageCodeISO639v2 = (code: string) => {
	return Boolean(
		code &&
			languages.some(
				(lang) => code === lang.v2 || code === lang.v2B || code === lang.v2T,
			),
	);
};

/**
 * Return subset of ISO-639 lang codes
 */
export const getLanguageCodesISO639 = (set: 'v1' | 'v2') => {
	const pickedLanguages: string[] = [];
	for (const lang of languages) {
		switch (set) {
			case 'v1':
				if (lang.v1) {
					pickedLanguages.push(lang.v1);
				}
				break;
			case 'v2':
				pickedLanguages.push(lang.v2);
				break;
		}
	}

	return pickedLanguages;
};
