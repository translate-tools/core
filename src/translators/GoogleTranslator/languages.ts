import { LanguageAliases } from '../../languages/LanguageAliases';

/**
 * Raw languages array
 */
// prettier-ignore
export const supportedLanguages = [
	'af', 'ak', 'am', 'ar', 'as', 'ay', 'az', 'be', 'bg', 'bho',
	'bm', 'bn', 'bs', 'ca', 'ceb', 'ckb', 'co', 'cs', 'cy', 'da',
	'de', 'doi', 'dv', 'ee', 'el', 'en', 'eo', 'es', 'et', 'eu',
	'fa', 'fi', 'fr', 'fy', 'ga', 'gd', 'gl', 'gn', 'gom', 'gu',
	'ha', 'haw', 'hi', 'hmn', 'hr', 'ht', 'hu', 'hy', 'id', 'ig',
	'ilo', 'is', 'it', 'iw', 'ja', 'jw', 'ka', 'kk', 'km', 'kn',
	'ko', 'kri', 'ku', 'ky', 'la', 'lb', 'lg', 'ln', 'lo', 'lt',
	'lus', 'lv', 'mai', 'mg', 'mi', 'mk', 'ml', 'mn', 'mni-Mtei', 'mr',
	'ms', 'mt', 'my', 'ne', 'nl', 'no', 'nso', 'ny', 'om', 'or',
	'pa', 'pl', 'ps', 'pt', 'qu', 'ro', 'ru', 'rw', 'sa', 'sd',
	'si', 'sk', 'sl', 'sm', 'sn', 'so', 'sq', 'sr', 'st', 'su',
	'sv', 'sw', 'ta', 'te', 'tg', 'th', 'ti', 'tk', 'tl', 'tr',
	'ts', 'tt', 'ug', 'uk', 'ur', 'uz', 'vi', 'xh', 'yi', 'yo',
	'zh', 'zh-CN', 'zh-TW', 'zu'
];

/**
 * Map with languages aliases.
 *
 * Google translator use legacy codes for some languages,
 * this map useful to use actual language codes by aliases
 *
 * @link https://xml.coverpages.org/iso639a.html
 */
export const fixedLanguagesMap: Record<string, string> = {
	he: 'iw',
	jv: 'jw',
};

/**
 * Map ISO lang codes to google translator lang codes
 */
export const languageAliases = new LanguageAliases(supportedLanguages, {
	map: fixedLanguagesMap,
});

/**
 * @param language language code or `auto`
 * @returns mapped language
 */
export const getFixedLanguage = (language: string) =>
	languageAliases.get(language) ?? language;
