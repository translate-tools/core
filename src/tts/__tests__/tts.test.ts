import { parseBuffer } from 'music-metadata';

import { isLanguageCodeISO639v1 } from '../../languages';

import { GoogleTTS } from '../GoogleTTS';
import { LingvaTTS } from '../LingvaTTS';
import { TTSProvider } from '..';

const ttsConstructor: TTSProvider[] = [GoogleTTS, LingvaTTS];

ttsConstructor.map((ttsConstructor) => {
	describe(`methods of TTS class "${ttsConstructor.name}"`, () => {
		test(`getAudioBuffer returns audio buffer`, async () => {
			const tts = new ttsConstructor();
			const audioBuffer = await tts.getAudioBuffer(
				'Hello world. This is a demo text from TTS module',
				'en',
			);

			expect(audioBuffer.buffer).toBeInstanceOf(ArrayBuffer);
			expect(typeof audioBuffer.type).toBe('string');

			// Parse as audio file
			const parsedAudioFrame = await parseBuffer(
				new Uint8Array(audioBuffer.buffer),
			);

			expect(parsedAudioFrame.format.bitrate).toBe(64000);
			expect(parsedAudioFrame.format.codec).toBe('MPEG 2 Layer 3');
			expect(parsedAudioFrame.format.container).toBe('MPEG');
			expect(parsedAudioFrame.format.duration).toBeGreaterThan(0);
			expect(parsedAudioFrame.format.hasAudio).toBe(true);
			expect(parsedAudioFrame.format.hasVideo).toBe(false);
			expect(parsedAudioFrame.format.sampleRate).toBe(24000);
			expect(parsedAudioFrame.format.numberOfSamples).toBeGreaterThan(100_000);
		});

		// Disable test, to allow TTS to return any lang codes they support
		// Users must filter lang codes on their side, to ensure valid ISO codes
		test.skip(`getSupportedLanguages returns array of supported languages`, async () => {
			const supportedLanguages = ttsConstructor.getSupportedLanguages();

			// Languages array are not empty
			expect(Array.isArray(supportedLanguages)).toBeTruthy();
			expect(supportedLanguages.length).not.toEqual(0);
			expect(supportedLanguages.length > 0).toBeTruthy();

			// All language coded are correct
			supportedLanguages.forEach((lang) => {
				expect(isLanguageCodeISO639v1(lang)).toBeTruthy();
			});
		});
	});
});
