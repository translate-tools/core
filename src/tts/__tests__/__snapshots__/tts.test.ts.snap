// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`methods of TTS class "GoogleTTS" > getAudioBuffer returns audio buffer 1`] = `
{
  "common": {
    "disk": {
      "no": null,
      "of": null,
    },
    "movementIndex": {
      "no": null,
      "of": null,
    },
    "track": {
      "no": null,
      "of": null,
    },
  },
  "format": {
    "bitrate": 64000,
    "codec": "MPEG 2 Layer 3",
    "codecProfile": "CBR",
    "container": "MPEG",
    "duration": 4.752,
    "hasAudio": true,
    "hasVideo": false,
    "lossless": false,
    "numberOfChannels": 1,
    "numberOfSamples": 114048,
    "sampleRate": 24000,
    "tagTypes": [],
    "trackInfo": [],
  },
  "native": {},
  "quality": {
    "warnings": [],
  },
}
`;

exports[`methods of TTS class "LingvaTTS" > getAudioBuffer returns audio buffer 1`] = `
{
  "common": {
    "disk": {
      "no": null,
      "of": null,
    },
    "movementIndex": {
      "no": null,
      "of": null,
    },
    "track": {
      "no": null,
      "of": null,
    },
  },
  "format": {
    "bitrate": 64000,
    "codec": "MPEG 2 Layer 3",
    "codecProfile": "CBR",
    "container": "MPEG",
    "duration": 4.56,
    "hasAudio": true,
    "hasVideo": false,
    "lossless": false,
    "numberOfChannels": 1,
    "numberOfSamples": 109440,
    "sampleRate": 24000,
    "tagTypes": [],
    "trackInfo": [],
  },
  "native": {},
  "quality": {
    "warnings": [],
  },
}
`;
